File: sdk/python/shennian_agent/agent.py
from __future__ import annotations

import json
import sys
import uuid
from collections.abc import Callable
from typing import Any, Optional

from .events import (
    DeltaEvent,
    ErrorEvent,
    FinalEvent,
    NotifyEvent,
    ToolCallEvent,
    ToolResultEvent,
    serialize,
)

SendHandler = Callable[[str, Optional[str], list[dict[str, str]]], None]
ResumeHandler = Callable[[str, str], None]



class Agent:
    def __init__(
        self,
        name: str,
        model: str,
        mode: str = "stdio",
        proactive: bool = False,
        resume: bool = False,
        version: str = "1.0.0",
    ) -> None:
        self.name = name
        self.model = model
        self.mode = mode
        self.proactive = proactive
        self._resume = resume
        self.version = version

        self._send_handler: SendHandler | None = None
        self._resume_handler: ResumeHandler | None = None
        self._run_id: str | None = None
        self._seq: int = 0

    def on_send(self, fn: SendHandler) -> SendHandler:
        self._send_handler = fn
        return fn

    def on_resume(self, fn: ResumeHandler) -> ResumeHandler:
        self._resume_handler = fn
        return fn

    def _new_run(self) -> None:
        self._run_id = str(uuid.uuid4())
        self._seq = 0

    def _emit(self, event_json: str) -> None:
        sys.stdout.write(event_json + "\n")
        sys.stdout.flush()

    def _next_seq(self) -> int:
        self._seq += 1
        return self._seq

    def delta(self, text: str, thinking: bool = False) -> None:
        if self._run_id is None:
            self._new_run()
        evt = DeltaEvent(text=text, run_id=self._run_id, seq=self._next_seq())
        if thinking:
            evt.thinking = True
        self._emit(serialize(evt))

    def tool_call(self, name: str, args: dict[str, Any] | None = None) -> None:
        if self._run_id is None:
            self._new_run()
        evt = ToolCallEvent(name=name, args=args, run_id=self._run_id, seq=self._next_seq())
        self._emit(serialize(evt))

    def tool_result(self, name: str, result: str) -> None:
        if self._run_id is None:
            self._new_run()
        evt = ToolResultEvent(name=name, result=result, run_id=self._run_id, seq=self._next_seq())
        self._emit(serialize(evt))

    def final(self, usage: dict[str, int] | None = None) -> None:
        if self._run_id is None:
            self._new_run()
        evt = FinalEvent(usage=usage, run_id=self._run_id, seq=self._next_seq())
        self._emit(serialize(evt))
        self._run_id = None
        self._seq = 0

    def error(self, message: str) -> None:
        if self._run_id is None:
            self._new_run()
        evt = ErrorEvent(message=message, run_id=self._run_id, seq=self._next_seq())
        self._emit(serialize(evt))
        self._run_id = None
        self._seq = 0

    def notify(self, text: str, title: str | None = None, source: str | None = None) -> None:
        evt = NotifyEvent(text=text, title=title, source=source)
        self._emit(serialize(evt))

    def run(self) -> None:
        if len(sys.argv) < 2:
            print(f"Usage: {sys.argv[0]} /caps | /run | /start", file=sys.stderr)
            sys.exit(1)

        cmd = sys.argv[1]
        if cmd == "/caps":
            self._handle_caps()
        elif cmd == "/run":
            self._handle_run()
        elif cmd == "/start":
            self._handle_start()
        else:
            print(f"Unknown command: {cmd}", file=sys.stderr)
            sys.exit(1)

    def _handle_caps(self) -> None:
        caps: dict[str, Any] = {
            "name": self.name,
            "model": self.model,
            "mode": self.mode,
            "version": self.version,
        }
        if self.proactive:
            caps["proactive"] = True
        if self._resume:
            caps["resume"] = True
        print(json.dumps(caps))

    def _handle_run(self) -> None:
        args = _parse_argv(sys.argv[2:])
        text = sys.stdin.read()
        self._new_run()
        if self._send_handler is None:
            self.error("no handler registered")
            return
        self._send_handler(text, args.get("session"), args.get("attachments", []))

    def _handle_start(self) -> None:
        for line in sys.stdin:
            line = line.strip()
            if not line:
                continue
            try:
                msg = json.loads(line)
            except json.JSONDecodeError:
                continue

            method = msg.get("method")
            params = msg.get("params", {})

            if method == "send":
                self._new_run()
                if self._send_handler is None:
                    self.error("no handler registered")
                    continue
                attachments = params.get("attachments", [])
                self._send_handler(params.get("text", ""), params.get("sessionId"), attachments)

            elif method == "resume":
                if self._resume_handler is None:
                    self._new_run()
                    self.error("resume not supported")
                    continue
                self._new_run()
                self._resume_handler(params.get("sessionId", ""), params.get("agentSessionId", ""))

            elif method == "stop":
                break


def _parse_argv(argv: list[str]) -> dict[str, Any]:
    result: dict[str, Any] = {"attachments": []}
    i = 0
    while i < len(argv):
        arg = argv[i]
        if arg.startswith("--workdir="):
            result["workdir"] = arg.split("=", 1)[1]
        elif arg == "--workdir" and i + 1 < len(argv):
            i += 1
            result["workdir"] = argv[i]
        elif arg.startswith("--session="):
            result["session"] = arg.split("=", 1)[1]
        elif arg == "--session" and i + 1 < len(argv):
            i += 1
            result["session"] = argv[i]
        elif arg.startswith("--resume="):
            result["resume"] = arg.split("=", 1)[1]
        elif arg == "--resume" and i + 1 < len(argv):
            i += 1
            result["resume"] = argv[i]
        elif arg.startswith("--attachment="):
            result["attachments"].append(arg.split("=", 1)[1])
        elif arg == "--attachment" and i + 1 < len(argv):
            i += 1
            result["attachments"].append(argv[i])
        i += 1
    return result

File: sdk/node/src/agent.ts
import { createInterface } from 'node:readline'
import { randomUUID } from 'node:crypto'
import type {
  AgentEvent,
  BaseFields,
  Caps,
  SendParams,
  ResumeParams,
  StdinMessage,
} from './events.js'

export type AgentOptions = {
  name: string
  model?: string
  models?: string[]
  defaultModel?: string
  mode?: 'spawn' | 'stdio'
  proactive?: boolean
  resume?: boolean
  version?: string
}

export class Agent {
  private caps: Caps
  private runId: string = ''
  private seq: number = 0
  private sendHandler?: (params: SendParams) => void | Promise<void>
  private resumeHandler?: (params: ResumeParams) => void | Promise<void>

  constructor(opts: AgentOptions) {
    const models = opts.models?.length
      ? opts.models
      : opts.model
        ? [opts.model]
        : []
    const defaultModel = opts.defaultModel ?? opts.model ?? models[0] ?? 'unknown'
    this.caps = {
      name: opts.name,
      model: defaultModel,
      models: models.length > 0 ? models : undefined,
      defaultModel: models.length > 0 ? defaultModel : undefined,
      mode: opts.mode ?? 'spawn',
      proactive: opts.proactive,
      resume: opts.resume,
      version: opts.version,
    }
  }

  onSend(handler: (params: SendParams) => void | Promise<void>): void {
    this.sendHandler = handler
  }

  onResume(handler: (params: ResumeParams) => void | Promise<void>): void {
    this.resumeHandler = handler
  }

  delta(text: string, thinking?: boolean): void {
    const event: AgentEvent = thinking
      ? { state: 'delta', text, thinking }
      : { state: 'delta', text }
    this.emit(event)
  }

  toolCall(name: string, args?: Record<string, unknown>): void {
    const event: AgentEvent = args
      ? { state: 'tool-call', name, args }
      : { state: 'tool-call', name }
    this.emit(event)
  }

  toolResult(name: string, result: string): void {
    this.emit({ state: 'tool-result', name, result })
  }

  final(usage?: { inputTokens: number; outputTokens: number }): void {
    const event: AgentEvent = usage
      ? { state: 'final', usage }
      : { state: 'final' }
    this.emit(event)
    this.resetRun()
  }

  error(message: string): void {
    this.emit({ state: 'error', message })
    this.resetRun()
  }

  notify(text: string, title?: string, source?: string): void {
    const event: Record<string, unknown> = { state: 'notify', text }
    if (title !== undefined) event.title = title
    if (source !== undefined) event.source = source
    this.emit(event as AgentEvent)
  }

  run(): void {
    const command = process.argv[2]

    if (command === '/caps') {
      this.writeLine(JSON.stringify(this.caps))
      process.exit(0)
    }

    if (command === '/run') {
      this.handleSpawnRun()
      return
    }

    if (command === '/start') {
      this.handleStdioMode()
      return
    }

    process.stderr.write(`Unknown command: ${command}\nUsage: <agent> /caps | /run | /start\n`)
    process.exit(1)
  }

  private emit(event: AgentEvent): void {
    if (!this.runId) this.newRun()
    const line: AgentEvent & BaseFields = { ...event, runId: this.runId, seq: this.seq++ }
    this.writeLine(JSON.stringify(line))
  }

  private writeLine(data: string): void {
    process.stdout.write(data + '\n')
  }

  private newRun(): void {
    this.runId = randomUUID()
    this.seq = 0
  }

  private resetRun(): void {
    this.runId = ''
    this.seq = 0
  }

  private handleSpawnRun(): void {
    const args = process.argv.slice(3)
    let sessionId = ''
    let resumeSessionId = ''
    let modelId = ''
    const attachments: { path: string; name: string; mime: string }[] = []

    for (let i = 0; i < args.length; i++) {
      switch (args[i]) {
        case '--session':
          sessionId = args[++i] ?? ''
          break
        case '--resume':
          resumeSessionId = args[++i] ?? ''
          break
        case '--workdir':
          process.chdir(args[++i] ?? '.')
          break
        case '--model':
          modelId = args[++i] ?? ''
          break
        case '--attachment': {
          const raw = args[++i] ?? ''
          const [path = '', name = '', mime = ''] = raw.split(':')
          attachments.push({ path, name, mime })
          break
        }
      }
    }

    if (resumeSessionId && this.resumeHandler) {
      this.newRun()
      void Promise.resolve(
        this.resumeHandler({ sessionId, agentSessionId: resumeSessionId })
      ).catch((err: Error) => this.error(err.message))
      return
    }

    const chunks: Buffer[] = []
    process.stdin.on('data', (chunk: Buffer) => chunks.push(chunk))
    process.stdin.on('end', () => {
      const text = Buffer.concat(chunks).toString('utf-8').trim()
      if (!this.sendHandler) return
      this.newRun()
      void Promise.resolve(
        this.sendHandler({ text, sessionId, attachments, modelId: modelId || undefined })
      ).catch((err: Error) => this.error(err.message))
    })
  }

  private handleStdioMode(): void {
    const rl = createInterface({ input: process.stdin })

    rl.on('line', (line) => {
      if (!line.trim()) return
      let msg: StdinMessage
      try {
        msg = JSON.parse(line) as StdinMessage
      } catch {
        return
      }

      switch (msg.method) {
        case 'send':
          if (this.sendHandler) {
            this.newRun()
            void Promise.resolve(
              this.sendHandler(msg.params)
            ).catch((err: Error) => this.error(err.message))
          }
          break
        case 'resume':
          if (this.resumeHandler) {
            this.newRun()
            void Promise.resolve(
              this.resumeHandler(msg.params)
            ).catch((err: Error) => this.error(err.message))
          }
          break
        case 'stop':
          rl.close()
          process.exit(0)
          break
      }
    })

    rl.on('close', () => process.exit(0))
  }
}
